zip-0222.rst

```diff
@@ -0,0 +1,203 @@
+  Title: Transparent Zcash Extensions
+  Owners: Jack Grigg <jack@electriccoin.co>
+  Credits: Zaki Manian <zaki@manian.org>
+           Daira Hopwood <daira@electriccoin.co>
+           Sean Bowe <sean@electriccoin.co>
+  Status: Draft
+  Category: Consensus
+  Created: 2019-07-01
+  License: MIT
+
+
+Terminology
+===========
+
+The key words "MUST" and "MAY" in this document are to be interpreted as described in
+RFC 2119. [#RFC2119]_
+
+The term "network upgrade" in this document is to be interpreted as described in ZIP 200
+[#zip-0200]_.
+
+The term "Sapling" in this document is to be interpreted as described in ZIP 205
+[#zip-0205]_.
+
+Abstract
+========
+
+This proposal defines a modification to the consensus rules that enables complex forms of
+transparent output predicates to be deployed in network upgrades. This in turn enables the
+creation of "transparent Zcash extensions" that augment the network's functionality in a
+carefully-defined and auditable fashion.
+
+
+Motivation
+==========
+
+Zcash transactions have three types of outputs: Sapling shielded outputs, Sprout shielded
+outputs, and transparent outputs. All three can by design be "encumbered" by a spending
+key - a user creating a transaction that spends the output must prove that they have
+access to the spending key controlling the output. For Sprout shielded outputs this was
+part of the zero-knowledge proof; for transparent and Sapling shielded outputs, the user
+creates a signature with the spending key.
+
+In some situations, more complex predicates are required. Zcash does not yet have any form
+of shielded programmability, and thus users rely on the script system inherited from
+Bitcoin to encumber transparent outputs. But even this is insufficient as-is for some
+desired use cases. In particular, the Bolt protocol [#zip-draft-bolt]_ requires a custom
+opcode, because the script system does not have the necessary primitives; this requires a
+change to the consensus rules.
+
+Shielded programmability would require significant changes across the ecosystem. In the
+interim, we propose in this ZIP a transparent form of Zcash extensibility that could later
+be migrated to shielded programmability.
+
+
+Conventions
+===========
+
+We use the following notation as defined in the Zcash protocol specification
+[#spec-notation]_, reproduced here for convenience:
+
+- length(*S*) means the length of (number of elements in) *S*.
+
+- { *a* .. *b* } means the set or type of integers from *a* through *b* inclusive.
+
+- *a* || *b* means the concatenation of sequences *a* then *b*.
+
+We reproduce the following conversion function from ZIP 32 [#zip-0032]_:
+
+- I2LEOSP\ :sub:`l`\ (*k*) is the byte sequence *S* of length *l*/8 representing in little-endian order the
+  integer *k* in range {0..2\ :sup:`l`\ -1}.
+
+
+Specification
+=============
+
+Transparent extension
+---------------------
+
+A transparent extension is identified by a numeric ``type``. A transparent extension may
```

Comment:
`type` is an odd name to use here - the "type" is just a unique numeric identifier, correct? Also, this seems like an unusual thing to start with - the "type" is just used to identify which extension handler is dispatched to.